File: code/packages/ruby/directed_graph/lib/coding_adventures/directed_graph/labeled_graph.rb
# frozen_string_literal: true

# --------------------------------------------------------------------------
# labeled_graph.rb — A directed graph where edges carry labels
# --------------------------------------------------------------------------
#
# == What is a labeled graph?
#
# A regular directed graph says "there IS an edge from A to B."  A labeled
# directed graph says "there is an edge from A to B *with label L*."
# Multiple labels can exist on the same pair of nodes, and the same label
# can appear on different edges.
#
# == Why do we need this?
#
# Labeled graphs are essential for modeling systems where the *kind* of
# relationship matters, not just its existence.  The canonical example is
# a finite state machine (FSM):
#
#   - Nodes are states (e.g. "locked", "unlocked")
#   - Edge labels are input symbols (e.g. "coin", "push")
#   - The edge ("locked", "unlocked", "coin") means:
#     "when in state 'locked' and input is 'coin', go to 'unlocked'"
#
# Self-loops are allowed because an FSM state can transition back to
# itself (e.g. inserting a coin in an already-unlocked turnstile).
#
# == Design: composition over inheritance
#
# Rather than inheriting from Graph, we *wrap* a Graph instance.  This
# is the "composition" pattern — LabeledGraph HAS-A Graph rather than
# IS-A Graph.  We do this because:
#
#   1. We need to set `allow_self_loops: true` on the inner graph.
#   2. The edge semantics are different — we store (from, to, label)
#      triples, not just (from, to) pairs.
#   3. We can still delegate algorithms like topological_sort directly
#      to the inner graph, since those only care about node connectivity.
#
# == Internal storage
#
# The labels are stored in a Hash keyed by [from, to] pairs:
#
#   @labels = { ["locked", "unlocked"] => Set["coin"],
#               ["unlocked", "locked"] => Set["push"] }
#
# The inner Graph handles node storage and adjacency.  The labels Hash
# adds the extra dimension of *which* labels exist on each edge.
# --------------------------------------------------------------------------

module CodingAdventures
  module DirectedGraph
    class LabeledGraph
      # ----------------------------------------------------------------
      # Construction
      # ----------------------------------------------------------------

      # Creates a new, empty labeled directed graph.
      #
      # The inner graph allows self-loops because labeled graphs commonly
      # model state machines where a state can transition to itself.
      def initialize
        @graph = Graph.new(allow_self_loops: true)
        @labels = {}
      end

      # ----------------------------------------------------------------
      # Node operations — delegated to the inner graph
      # ----------------------------------------------------------------
      #
      # Nodes in a labeled graph work exactly like nodes in a regular
      # graph.  We simply forward these calls.

      # Adds a node to the graph.  No-op if it already exists.
      def add_node(node)
        @graph.add_node(node)
        self
      end

      # Removes a node and ALL edges (with all labels) touching it.
      #
      # We must clean up the labels Hash entries for any edge that
      # involves this node — both outgoing and incoming.
      #
      # Raises NodeNotFoundError if the node is not in the graph.
      def remove_node(node)
        unless has_node?(node)
          raise NodeNotFoundError, "Node not found: #{node.inspect}"
        end

        # Clean up label entries for outgoing edges.
        @graph.successors(node).each do |succ|
          @labels.delete([node, succ])
        end

        # Clean up label entries for incoming edges.
        @graph.predecessors(node).each do |pred|
          @labels.delete([pred, node])
        end

        # Clean up self-loop label entries if present.
        @labels.delete([node, node])

        @graph.remove_node(node)
        self
      end

      # Returns true if the node exists in the graph.
      def has_node?(node)
        @graph.has_node?(node)
      end

      # Returns a sorted array of all nodes.
      def nodes
        @graph.nodes
      end

      # Returns the number of nodes in the graph.
      def size
        @graph.size
      end

      # ----------------------------------------------------------------
      # Labeled edge operations
      # ----------------------------------------------------------------
      #
      # These are the methods that distinguish LabeledGraph from Graph.
      # Every edge operation requires a label.

      # Adds a labeled edge from +from_node+ to +to_node+ with +label+.
      #
      # Both nodes are implicitly created if they don't exist.
      # If the exact same (from, to, label) triple already exists, this
      # is a no-op (Set deduplication on the label set).
      #
      # Adding a second label to the same (from, to) pair creates a
      # multi-labeled edge — the underlying graph still has one structural
      # edge, but the labels Set grows.
      def add_edge(from_node, to_node, label)
        # Ensure the structural edge exists in the inner graph.
        # The inner graph handles node creation and allows self-loops.
        @graph.add_edge(from_node, to_node) unless @graph.has_edge?(from_node, to_node)

        key = [from_node, to_node]
        @labels[key] ||= Set.new
        @labels[key].add(label)
        self
      end

      # Removes a specific labeled edge.
      #
      # If this was the last label on the (from, to) pair, the
      # structural edge is also removed from the inner graph.
      #
      # Raises NodeNotFoundError if either node doesn't exist.
      # Raises EdgeNotFoundError if the (from, to, label) triple doesn't exist.
      def remove_edge(from_node, to_node, label)
        unless has_node?(from_node)
          raise NodeNotFoundError, "Node not found: #{from_node.inspect}"
        end
        unless has_node?(to_node)
          raise NodeNotFoundError, "Node not found: #{to_node.inspect}"
        end

        key = [from_node, to_node]
        unless @labels.key?(key) && @labels[key].include?(label)
          raise EdgeNotFoundError,
            "Edge not found: #{from_node.inspect} -> #{to_node.inspect} [#{label.inspect}]"
        end

        @labels[key].delete(label)

        # If no labels remain, remove the structural edge too.
        if @labels[key].empty?
          @labels.delete(key)
          @graph.remove_edge(from_node, to_node)
        end

        self
      end

      # Checks whether an edge exists.
      #
      # - With a label:  returns true if the exact (from, to, label) triple exists.
      # - Without a label: returns true if ANY edge from -> to exists (regardless of label).
      def has_edge?(from_node, to_node, label = nil)
        if label.nil?
          @graph.has_edge?(from_node, to_node)
        else
          key = [from_node, to_node]
          @labels.key?(key) && @labels[key].include?(label)
        end
      end

      # Returns all edges as [from, to, label] triples, sorted for
      # deterministic output.
      #
      # If a pair (A, B) has labels {"x", "y"}, we return two triples:
      #   ["A", "B", "x"] and ["A", "B", "y"]
      def edges
        result = []
        @labels.each do |(from_node, to_node), label_set|
          label_set.each do |label|
            result << [from_node, to_node, label]
          end
        end
        result.sort
      end

      # Returns the Set of labels on the edge from +from_node+ to +to_node+.
      #
      # Returns an empty Set if no edge exists between the two nodes.
      #
      # Raises NodeNotFoundError if either node doesn't exist.
      def labels(from_node, to_node)
        unless has_node?(from_node)
          raise NodeNotFoundError, "Node not found: #{from_node.inspect}"
        end
        unless has_node?(to_node)
          raise NodeNotFoundError, "Node not found: #{to_node.inspect}"
        end

        key = [from_node, to_node]
        @labels.key?(key) ? @labels[key].dup : Set.new
      end

      # ----------------------------------------------------------------
      # Neighbor queries with optional label filtering
      # ----------------------------------------------------------------
      #
      # These extend the basic Graph predecessors/successors with an
      # optional `label:` keyword that filters by edge label.

      # Returns successors of +node+, optionally filtered by +label+.
      #
      # - Without label: returns ALL direct successors (same as Graph).
      # - With label: returns only successors reachable via an edge
      #   carrying that specific label.
      #
      # Raises NodeNotFoundError if the node doesn't exist.
      def successors(node, label: nil)
        unless has_node?(node)
          raise NodeNotFoundError, "Node not found: #{node.inspect}"
        end

        if label.nil?
          @graph.successors(node)
        else
          result = []
          @graph.successors(node).each do |succ|
            key = [node, succ]
            result << succ if @labels.key?(key) && @labels[key].include?(label)
          end
          result.sort
        end
      end

      # Returns predecessors of +node+, optionally filtered by +label+.
      #
      # - Without label: returns ALL direct predecessors (same as Graph).
      # - With label: returns only predecessors connected via an edge
      #   carrying that specific label.
      #
      # Raises NodeNotFoundError if the node doesn't exist.
      def predecessors(node, label: nil)
        unless has_node?(node)
          raise NodeNotFoundError, "Node not found: #{node.inspect}"
        end

        if label.nil?
          @graph.predecessors(node)
        else
          result = []
          @graph.predecessors(node).each do |pred|
            key = [pred, node]
            result << pred if @labels.key?(key) && @labels[key].include?(label)
          end
          result.sort
        end
      end

      # ----------------------------------------------------------------
      # Algorithm delegation
      # ----------------------------------------------------------------
      #
      # Graph algorithms operate on the structural topology (which nodes
      # are connected to which), not on labels.  We delegate directly to
      # the inner Graph.

      # Returns a topological ordering of all nodes.
      # Raises CycleError if the graph contains a cycle.
      def topological_sort
        @graph.topological_sort
      end

      # Returns true if the graph contains at least one cycle.
      def has_cycle?
        @graph.has_cycle?
      end

      # Returns the full transitive closure as a Hash of Sets.
      def transitive_closure
        @graph.transitive_closure
      end
    end
  end
end

File: code/packages/lua/directed_graph/src/coding_adventures/directed_graph/labeled_graph.lua
-- labeled_graph — Labeled Directed Graph
-- =======================================
--
-- A LabeledGraph extends the basic directed graph with edge labels.
-- Each edge can have one or more string labels, turning the graph into
-- a "multigraph-like" structure where the same pair of nodes can be
-- connected by edges with different semantic meanings.
--
-- # Why labeled edges?
--
-- In a build system, you might want to distinguish between different
-- kinds of dependencies:
--
--   - "compile" dependency: package A needs package B at compile time
--   - "test"    dependency: package A needs package B only for testing
--   - "runtime" dependency: package A needs package B at runtime
--
-- With labeled edges, you can query "what are A's compile-time
-- dependencies?" without conflating them with test-only dependencies.
--
-- # Architecture: composition over inheritance
--
-- Rather than duplicating the adjacency-map logic from DirectedGraph,
-- LabeledGraph wraps a DirectedGraph and adds a label map on top:
--
--     +---------------------------------------------------+
--     | LabeledGraph                                      |
--     |                                                   |
--     |  +-------------------------+                      |
--     |  | graph (DirectedGraph)   |  <- handles nodes,   |
--     |  |   forward adjacency     |    edges, algorithms  |
--     |  |   reverse adjacency     |                      |
--     |  +-------------------------+                      |
--     |                                                   |
--     |  labels["from\0to"] = { label1=true, label2=true } |
--     |    (from, to) -> set of label strings             |
--     |                                                   |
--     +---------------------------------------------------+
--
-- The underlying DirectedGraph stores a single edge between any two
-- nodes, regardless of how many labels that edge carries. When all
-- labels for an edge are removed, the underlying edge is also removed.
--
-- All algorithm methods (topological_sort, has_cycle, transitive_closure,
-- etc.) delegate to the underlying DirectedGraph, so they work identically.
--
-- # Label map key encoding
--
-- In Go, the label map key is [2]string{from, to} — a fixed-size array.
-- Lua tables can't use arrays as keys (they'd be compared by reference,
-- not value). Instead, we concatenate from and to with a null byte
-- separator: from .. "\0" .. to. The null byte is safe because node names
-- are strings that shouldn't contain null bytes.

-- Helper: make a label map key from (from, to) pair.
-- We use null-byte separation to avoid collisions.
-- E.g., ("A", "B") -> "A\0B"
local function label_key(from, to)
    return from .. "\0" .. to
end

-- Error constructors — duplicated from init.lua to avoid circular require.
-- These are lightweight table constructors, so duplication is acceptable.

local function NodeNotFoundError(node)
    return {
        type = "node_not_found",
        node = node,
        message = string.format("node not found: %q", node),
    }
end

local function EdgeNotFoundError(from, to)
    return {
        type = "edge_not_found",
        from = from,
        to = to,
        message = string.format("edge not found: %q -> %q", from, to),
    }
end

local function LabelNotFoundError(from, to, label)
    return {
        type = "label_not_found",
        from = from,
        to = to,
        label = label,
        message = string.format("label %q not found on edge %q -> %q", label, from, to),
    }
end

-- =========================================================================
-- LabeledGraph class
-- =========================================================================

local LabeledGraph = {}
LabeledGraph.__index = LabeledGraph

--- Create an empty labeled directed graph.
-- Self-loops are prohibited (inherits from DirectedGraph.new()).
--
-- @return LabeledGraph A new empty labeled graph.
function LabeledGraph.new()
    local dg = require("coding_adventures.directed_graph")
    local self = setmetatable({}, LabeledGraph)
    self._graph = dg.DirectedGraph.new()
    self._labels = {}  -- label_key(from,to) -> {label1=true, label2=true}
    return self
end

--- Create an empty labeled directed graph that permits self-loops.
--
-- @return LabeledGraph A new empty labeled graph allowing self-loops.
function LabeledGraph.new_allow_self_loops()
    local dg = require("coding_adventures.directed_graph")
    local self = setmetatable({}, LabeledGraph)
    self._graph = dg.DirectedGraph.new_allow_self_loops()
    self._labels = {}
    return self
end

-- =========================================================================
-- Node operations — delegate to underlying DirectedGraph
-- =========================================================================

--- Add a node to the graph. No-op if the node already exists.
-- @param node string The node identifier.
function LabeledGraph:add_node(node)
    self._graph:add_node(node)
end

--- Remove a node and all its incident edges (including labels).
--
-- When a node is removed, we must also clean up any label entries for
-- edges that touched that node. We iterate over all label keys and
-- remove any that reference the deleted node.
--
-- @param node string The node to remove.
-- @return true on success, or nil + error.
function LabeledGraph:remove_node(node)
    if not self._graph:has_node(node) then
        return nil, NodeNotFoundError(node)
    end

    -- Clean up labels for all edges involving this node.
    -- We collect keys first to avoid mutating the table during iteration.
    local keys_to_delete = {}
    for key, _ in pairs(self._labels) do
        -- Key format is "from\0to". Check if either part is our node.
        local sep = key:find("\0", 1, true)
        local from = key:sub(1, sep - 1)
        local to = key:sub(sep + 1)
        if from == node or to == node then
            keys_to_delete[#keys_to_delete + 1] = key
        end
    end
    for _, key in ipairs(keys_to_delete) do
        self._labels[key] = nil
    end

    return self._graph:remove_node(node)
end

--- Check whether a node exists in the graph.
-- @param node string The node to check.
-- @return boolean
function LabeledGraph:has_node(node)
    return self._graph:has_node(node)
end

--- Return all nodes in sorted order (deterministic).
-- @return table A sorted array of node strings.
function LabeledGraph:nodes()
    return self._graph:nodes()
end

--- Return the number of nodes in the graph.
-- @return integer
function LabeledGraph:size()
    return self._graph:size()
end

-- =========================================================================
-- Labeled edge operations
-- =========================================================================
--
-- Each edge in a LabeledGraph carries one or more labels. add_edge
-- requires a label; if you want multiple labels on the same edge,
-- call add_edge multiple times with different labels.
--
-- The underlying DirectedGraph tracks whether an edge exists at all (for
-- algorithm purposes). The label map tracks which labels are on each edge.

--- Add a directed edge from `from` to `to` with the given label.
--
-- If the edge already exists (possibly with different labels), the new
-- label is added to the existing set — the edge is not duplicated in
-- the underlying graph.
--
-- If the edge does not yet exist, it is created in the underlying graph
-- and the label is recorded.
--
-- Raises an error on self-loops if the underlying graph prohibits them.
--
-- @param from string The source node.
-- @param to string The destination node.
-- @param label string The label for this edge.
function LabeledGraph:add_edge(from, to, label)
    -- This may raise an error if self-loops are not allowed.
    self._graph:add_edge(from, to)

    local key = label_key(from, to)
    if not self._labels[key] then
        self._labels[key] = {}
    end
    self._labels[key][label] = true
end

--- Remove a specific label from the edge from->to.
--
-- If this was the last label on the edge, the underlying edge is also
-- removed from the graph. If other labels remain, only the specified
-- label is removed and the edge persists.
--
-- @param from string The source node.
-- @param to string The destination node.
-- @param label string The label to remove.
-- @return true on success, or nil + error.
function LabeledGraph:remove_edge(from, to, label)
    local key = label_key(from, to)
    local label_set = self._labels[key]

    if not label_set or not self._graph:has_edge(from, to) then
        return nil, EdgeNotFoundError(from, to)
    end

    if not label_set[label] then
        return nil, LabelNotFoundError(from, to, label)
    end

    label_set[label] = nil

    -- If no labels remain, remove the underlying edge entirely.
    local remaining = false
    for _ in pairs(label_set) do
        remaining = true
        break
    end

    if not remaining then
        self._labels[key] = nil
        return self._graph:remove_edge(from, to)
    end

    return true
end

--- Check whether there's any edge from `from` to `to` (regardless of label).
-- @param from string The source node.
-- @param to string The destination node.
-- @return boolean
function LabeledGraph:has_edge(from, to)
    return self._graph:has_edge(from, to)
end

--- Check whether there's an edge from `from` to `to` with a specific label.
-- @param from string The source node.
-- @param to string The destination node.
-- @param label string The label to check for.
-- @return boolean
function LabeledGraph:has_edge_with_label(from, to, label)
    local key = label_key(from, to)
    local label_set = self._labels[key]
    if label_set then
        return label_set[label] == true
    end
    return false
end

--- Return all edges as {from, to, label} triples, sorted deterministically.
--
-- If an edge has multiple labels, it appears once per label.
--
-- Example: if edge A->B has labels "compile" and "test", the output
-- includes both {"A", "B", "compile"} and {"A", "B", "test"}.
--
-- @return table An array of {from, to, label} arrays.
function LabeledGraph:edges()
    local result = {}
    for key, label_set in pairs(self._labels) do
        local sep = key:find("\0", 1, true)
        local from = key:sub(1, sep - 1)
        local to = key:sub(sep + 1)
        for label, _ in pairs(label_set) do
            result[#result + 1] = {from, to, label}
        end
    end
    table.sort(result, function(a, b)
        if a[1] ~= b[1] then return a[1] < b[1] end
        if a[2] ~= b[2] then return a[2] < b[2] end
        return a[3] < b[3]
    end)
    return result
end

--- Return the set of labels on the edge from->to.
--
-- Returns an empty table if the edge doesn't exist. Returns a COPY
-- of the internal label set to prevent callers from mutating internal state.
--
-- @param from string The source node.
-- @param to string The destination node.
-- @return table A set (table with boolean values) of labels.
function LabeledGraph:labels(from, to)
    local key = label_key(from, to)
    local label_set = self._labels[key]
    if label_set then
        -- Return a copy to prevent callers from mutating internal state.
        local result = {}
        for l, _ in pairs(label_set) do
            result[l] = true
        end
        return result
    end
    return {}
end

-- =========================================================================
-- Neighbor queries with label filtering
-- =========================================================================

--- Return the direct successors of a node (any label).
-- @param node string The node to query.
-- @return table Sorted array of successor strings, or nil + error.
function LabeledGraph:successors(node)
    return self._graph:successors(node)
end

--- Return successors connected by edges with the given label.
--
-- For example, if A->B has label "compile" and A->C has label "test",
-- then successors_with_label("A", "compile") returns {"B"}.
--
-- @param node string The node to query.
-- @param label string The label to filter by.
-- @return table Sorted array of successor strings, or nil + error.
function LabeledGraph:successors_with_label(node, label)
    if not self._graph:has_node(node) then
        return nil, NodeNotFoundError(node)
    end

    local succs, err = self._graph:successors(node)
    if not succs then return nil, err end

    local result = {}
    for _, succ in ipairs(succs) do
        local key = label_key(node, succ)
        local label_set = self._labels[key]
        if label_set and label_set[label] then
            result[#result + 1] = succ
        end
    end
    table.sort(result)
    return result
end

--- Return the direct predecessors of a node (any label).
-- @param node string The node to query.
-- @return table Sorted array of predecessor strings, or nil + error.
function LabeledGraph:predecessors(node)
    return self._graph:predecessors(node)
end

--- Return predecessors connected by edges with the given label.
--
-- @param node string The node to query.
-- @param label string The label to filter by.
-- @return table Sorted array of predecessor strings, or nil + error.
function LabeledGraph:predecessors_with_label(node, label)
    if not self._graph:has_node(node) then
        return nil, NodeNotFoundError(node)
    end

    local preds, err = self._graph:predecessors(node)
    if not preds then return nil, err end

    local result = {}
    for _, pred in ipairs(preds) do
        local key = label_key(pred, node)
        local label_set = self._labels[key]
        if label_set and label_set[label] then
            result[#result + 1] = pred
        end
    end
    table.sort(result)
    return result
end

-- =========================================================================
-- Algorithm delegation
-- =========================================================================
--
-- All graph algorithms delegate to the underlying DirectedGraph. Labels
-- don't affect the structural algorithms — topological sort, cycle
-- detection, and transitive closure only care about whether edges exist,
-- not what they're labeled.

--- Return nodes in topological order.
-- @return table Sorted array of nodes, or nil + CycleError.
function LabeledGraph:topological_sort()
    return self._graph:topological_sort()
end

--- Return true if the graph contains a cycle.
-- @return boolean
function LabeledGraph:has_cycle()
    return self._graph:has_cycle()
end

--- Return all nodes reachable from the given node by following edges forward.
-- @param node string The starting node.
-- @return table A set of reachable nodes, or nil + error.
function LabeledGraph:transitive_closure(node)
    return self._graph:transitive_closure(node)
end

--- Return the underlying DirectedGraph, giving access to all base graph
-- methods (independent_groups, affected_nodes, etc.).
-- @return DirectedGraph The underlying graph.
function LabeledGraph:graph()
    return self._graph
end

return LabeledGraph
